### PR TITLE
Manifest resource detector

### LINKED ISSUE #10615: detect service.name and service.version based on `Implementation-Title` and `Implementation-Version`
### Is your feature request related to a problem? Please describe.

- produces better results than jar file detector
- is not limited to spring boot, where a specialized detector based on build metadata exists: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10515

### Describe the solution you'd like

- see https://stackoverflow.com/questions/921667/how-do-i-add-an-implementation-version-value-to-a-jar-manifest-using-maven
- possible implementation: https://github.com/grafana/grafana-opentelemetry-starter/blob/main/src/main/java/com/grafana/opentelemetry/OpenTelemetryConfig.java#L211-L232

### Gradle

```groovy

jar {
    manifest {
        attributes("Implementation-Title": project.name, "Implementation-Version": version)
    }
}
```

### PR DESCRIPTION
**Pull request by zeitlinger**
Fixes #10615

### CODE REVIEW

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/resources/OtelResourceAutoConfiguration.java [hunk 1]
```diff
@@ -48,6 +49,12 @@ public ResourceProvider otelDistroVersionResourceProvider() {
     return new DistroVersionResourceProvider();
   }
 
+  @Bean
+  @ConditionalOnClass(ManifestResourceProvider.class)
+  public ResourceProvider otelJarNameResourceProvider() {
```
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/resources/OtelResourceAutoConfiguration.java:54]:
```suggestion
  public ResourceProvider otelManifestResourceProvider() {
```

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  // Visible for testing
+  ManifestResourceProvider(SystemHelper systemHelper) {
+    super(
+        new AttributeProvider<Manifest>() {
+          @Override
+          public Optional<Manifest> readData() {
+            return Optional.ofNullable(
+                    systemHelper.openClasspathResource("META-INF", "MANIFEST.MF"))
```
Comment (trask) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:38]: which `MANIFEST.MF` will this get when dependencies are directly on the class path?

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeProvider.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * An easier alternative to {@link io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider}, which
+ * avoids some common pitfalls and boilerplate.
+ *
+ * <p>An example of how to use this interface can be found in {@link ManifestResourceProvider}.
+ */
+public interface AttributeProvider<D> {
```
Comment (jeanbisutti) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeProvider.java:19]: Should it be an internal class?

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarFileDetector.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.jar.Manifest;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+class JarFileDetector {
```
Comment (jeanbisutti) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarFileDetector.java:24]: `SpringBootServiceNameDetector `and  `JarServiceNameDetector ` implement `ConditionalResourceProvider`, which is not the case for `JarFileDetector`. So, the name `JarFileDetector` is a bit confusing.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.resources;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+public class ResourceDiscoveryCache {
```
Comment (jeanbisutti) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java:11]: Could it be package private?

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  private static final ConcurrentHashMap<String, Object> cache = new ConcurrentHashMap<>();
```
Comment (jeanbisutti) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java:12]: Today, the map seems to contain only one element when it is used in the implementation code. So, should it be a map?

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarFileDetector.java (hunk 4 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarFileDetector.java:24]: Right - will rename to `JarPathFinder`

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java (hunk 6 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java:12]: I mainly used it because it's a very easy pattern to get concurrency right.

Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ResourceDiscoveryCache.java]: removed

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+class JarPathFinder {
+  private final Supplier<String[]> getProcessHandleArguments;
+  private final Function<String, String> getSystemProperty;
+  private final Predicate<Path> fileExists;
+
+  private static Optional<Optional<Path>> jarPath = Optional.empty();
```
Comment (jeanbisutti) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java:23]: Would it be possible to have `Optional<Optional<Path>>` to simplify?

Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java:23]: no

- the outer optional is to know if detection has happened
- the inner optional is to know if a path has been found

Comment (jeanbisutti) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java:23]: > no
> 
> * the outer optional is to know if detection has happened
> * the inner optional is to know if a path has been found

This can't be quickly guess for me by reading the code.

So, `jarPath` seems to be related to 2 different things.

 Could you please update the code to make these things more explicit in the code, for example with another property related to the detection?

File: dependencyManagement/build.gradle.kts [hunk 8]
```diff
@@ -96,7 +97,9 @@ val DEPENDENCIES = listOf(
   "commons-logging:commons-logging:1.3.0",
   "commons-validator:commons-validator:1.8.0",
   "io.netty:netty:3.10.6.Final",
-  "io.opentelemetry.contrib:opentelemetry-aws-xray-propagator:1.33.0-alpha",
+  "io.opentelemetry.contrib:opentelemetry-aws-resources:${otelContribVersion}",
+  "io.opentelemetry.contrib:opentelemetry-aws-xray-propagator:${otelContribVersion}",
+  "io.opentelemetry.contrib:opentelemetry-gcp-resources:${otelContribVersion}",
```
Comment (trask) [inline dependencyManagement/build.gradle.kts:102]: if these changes aren't needed for the manifest resource detector can you move them into a separate PR?

Comment (zeitlinger) [reply dependencyManagement/build.gradle.kts]: That was a rebase gone wrong - thanks for spotting :smile:

**Review by jeanbisutti — APPROVED** [review]

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      return Optional.of(manifest);
+    } catch (Exception e) {
+      logger.log(WARNING, "Error reading manifest", e);
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public int order() {
+    // make it run later than SpringBootServiceNameDetector
```
Comment (trask) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:80]:
```suggestion
    // make it run later than SpringBootServiceNameDetector but before JarServiceNameDetector
```

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java
Comment (trask) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:1]: suggestion to update existing comment near bottom
```
    // make it run later than ManifestResourceProvider and SpringBootServiceNameDetector
```

Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:1]: done

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    @CanIgnoreReturnValue
+    @Override
+    public <T> AttributeBuilder add(AttributeKey<T> key, Function<D, Optional<T>> getter) {
+      attributeGetters.put((AttributeKey) key, Objects.requireNonNull((Function) getter));
+      return this;
+    }
+  }
+
+  private static final ThreadLocal<Resource> existingResource = new ThreadLocal<>();
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java:45]: I think using thread local isn't necessary here, could just use an instance variable.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+class JarPathFinder {
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarPathFinder.java:18]: perhaps this should be called `MainJarPathFinder` to make it more clear which jar it is looking for?

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+ * service.version</code> from META-INF/MANIFEST.MF.
+ */
+@AutoService(ResourceProvider.class)
+public final class ManifestResourceProvider extends AttributeResourceProvider<Manifest> {
+
+  private static final Logger logger = Logger.getLogger(ManifestResourceProvider.class.getName());
+
+  @SuppressWarnings("unused") // SPI
+  public ManifestResourceProvider() {
+    this(new JarPathFinder(), ManifestResourceProvider::readManifest);
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:32]: both this class and `JarServiceNameDetector` attempt to find the main jar, you could find it only once and use the result for both

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  private static Optional<Manifest> readManifest(Path jarPath) {
+    try (InputStream s =
+        new URL(String.format("jar:%s!/META-INF/MANIFEST.MF", jarPath.toUri())).openStream()) {
+      Manifest manifest = new Manifest();
+      manifest.read(s);
+      return Optional.of(manifest);
+    } catch (Exception e) {
+      logger.log(WARNING, "Error reading manifest", e);
+      return Optional.empty();
+    }
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:75]: I would have used
```java
    try (JarFile jarFile = new JarFile(path.toFile(), false)) {
      return Optional.of(jarFile.getManifest());
    } catch (IOException exception) {
      logger.log(WARNING, "Error reading manifest", exception);
      return Optional.empty();
    }
```

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * An easier alternative to {@link io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider}, which
+ * avoids some common pitfalls and boilerplate.
+ *
+ * <p>An example of how to use this interface can be found in {@link ManifestResourceProvider}.
+ */
+@SuppressWarnings({"unchecked", "rawtypes"})
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java:28]: I'd try to suppress the warnings only for the method that needs it

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java (hunk 10 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java:45]: There's one use case where we need it:

1. service name is already present in existing (because another provider has detected it), but version is not
2. should apply will return true (because version is not found)
3. `createResource` does not have access to `existing` - it's not passed as a param

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java (hunk 12 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:32]: we already do this: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10621/files#diff-9274d424baf0a4a58f5c616a81cf4ec659fce87b4e13e6f423414fe0e2a0602cR52-R57

Comment (laurit) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:32]: Didn't expect that. Instead of passing around `JarPathFinder` I would have passed around `Supplier<Optional<Path>>` I would have create a separate class `MainJarPathHolder` that has `private static final Optional<Path> jarPath = new JarPathFinder().getJarPath();` and used the jar path from the holder in both detectors.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java (hunk 10 above)
Comment (laurit) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java:45]: I don't get why this would require keeping the existing resource in a thread local, instance variable could be set in the `shouldApply` and used in `createResource` the same way as the thread local. Imo there isn't anything that would require that these classes have to be stateless. If you want you could figure out which attributes need to be set in `shouldApply` and keep that info as instance variable and use it in `createResource` then you probably wouldn't need the existing resource there at all.

Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java:45]: good idea!

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java (hunk 12 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:32]: good idea - let me try that

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java (hunk 10 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/AttributeResourceProvider.java:45]: Along the way, I also found a bug - so let me try to get that into the release

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java (hunk 12 above)
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ManifestResourceProvider.java:32]: I've done it a bit differently now, because the finder needs to be passed around in unit tests

**Comment by zeitlinger** [comment]
@laurit thanks for the review - please have a look now :smile:
